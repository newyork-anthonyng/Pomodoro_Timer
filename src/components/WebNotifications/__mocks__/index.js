import { PureComponent } from 'react';

class WebNotifications extends PureComponent {
  constructor() {
    super();

    this.createNotification = jest.fn();
  }

  render() {
    return null;
  }
}

export default WebNotifications;
