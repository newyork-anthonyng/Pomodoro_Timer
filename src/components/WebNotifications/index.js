import { PureComponent } from 'react';

class WebNotifications extends PureComponent {
  constructor() {
    super();

    this.notificationSupported = false;
  }

  componentDidMount() {
    if (!('Notification' in window)) {
      // eslint-disable-next-line no-console
      console.error('This browser does not support desktop notification');

    } else if (window.Notification.permission === 'granted') {
      this.notificationSupported = true;

    } else if (window.Notification.permission !== 'denied') {
      window.Notification.requestPermission().then((result) => {
        if (result === 'granted') {
          this.notificationSupported = true;
        }
      });
    }
  }

  createNotification({ title = 'Pomodoro Timer expired', body = '' } = {}) {
    if (!this.notificationSupported) return;

    const options = {
      body,
      requireInteraction: true,
    };
    new window.Notification(title, options);
  }

  render() {
    return null;
  }
}

export default WebNotifications;
