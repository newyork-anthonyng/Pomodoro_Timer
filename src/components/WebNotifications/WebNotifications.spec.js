import React from 'react';
import { mount } from 'enzyme';
import WebNotifications from './';

/*
 * Shhh...remove console from jest output
 */
 /* eslint-disable no-console */
console.log = jest.fn();
console.error = jest.fn();
/* eslint-enable no-console */

beforeEach(() => {
  delete window.Notification;
});

describe('Notification is not supported in window', () => {
  it('should do nothing', () => {
    const wrapper = mount(<WebNotifications />);

    expect(() => {
      wrapper.instance().createNotification();
    }).not.toThrow();
  });
});

describe('Notification permission is already granted', () => {
  beforeEach(() => {
    window.Notification = jest.fn();
    window.Notification.permission = 'granted';
  });

  it('should create notification without errors', () => {
    const wrapper = mount(<WebNotifications />);

    expect(() => {
      wrapper.instance().createNotification();
    }).not.toThrow();
    expect(window.Notification).toHaveBeenCalledTimes(1);
  });

  it('should create notification with correct defaults', () => {
    const wrapper = mount(<WebNotifications />);

    wrapper.instance().createNotification();

    expect(window.Notification.mock.calls[0]).toMatchSnapshot();
  });

  it('should pass correct options when creating nofications', () => {
    const wrapper = mount(<WebNotifications />);

    wrapper.instance().createNotification({
      title: 'Batman v Superman',
      body: 'Best movie ever',
    });

    expect(window.Notification.mock.calls[0]).toMatchSnapshot();
  });
});

describe('Notification permission was not denied', () => {
  let requestPermissionPromise;

  beforeEach(() => {
    window.Notification = jest.fn();
    window.Notification.permission = 'something else';

    window.Notification.requestPermission = jest.fn(() => {
      requestPermissionPromise = Promise.resolve();
      return requestPermissionPromise;
    });
  });

  it('should request permission', () => {
    mount(<WebNotifications />);

    expect(window.Notification.requestPermission).toHaveBeenCalledTimes(1);
  });

  it('should create notification without errors, if permission was granted', () => {
    window.Notification.requestPermission = jest.fn(() => {
      return Promise.resolve('granted');
    });
    const wrapper = mount(<WebNotifications />);

    return requestPermissionPromise.then(() => {
      expect(() => {
        wrapper.instance().createNotification();
      }).not.toThrow();
      expect(window.Notification).toHaveBeenCalledTimes(1);
    });
  });

  it('should do nothing if permission was not granted', () => {
    window.Notification.requestPermission = jest.fn(() => {
      return Promise.resolve('something else');
    });
    const wrapper = mount(<WebNotifications />);

    return requestPermissionPromise.then(() => {
      expect(() => {
        wrapper.instance().createNotification();
      }).not.toThrow();
      expect(window.Notification).not.toHaveBeenCalled();
    });
  });
});

describe('Notification permission was denied', () => {
  it('should do nothing', () => {
    window.Notification = jest.fn();
    window.Notification.permission = 'denied';
    const wrapper = mount(<WebNotifications />);

    expect(() => {
      wrapper.instance().createNotification();
    }).not.toThrow();
    expect(window.Notification).not.toHaveBeenCalled();
  });
});
