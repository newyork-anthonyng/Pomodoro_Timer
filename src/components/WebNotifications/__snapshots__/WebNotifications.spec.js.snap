// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification permission is already granted should create notification with correct defaults 1`] = `
Array [
  "Pomodoro Timer expired",
  Object {
    "body": "",
    "requireInteraction": true,
  },
]
`;

exports[`Notification permission is already granted should pass correct options when creating nofications 1`] = `
Array [
  "Batman v Superman",
  Object {
    "body": "Best movie ever",
    "requireInteraction": true,
  },
]
`;
